File: strategy/base/vis.lua
--[[
--- Provides functions to draw on the game field
module "vis"
]]--

--[[***********************************************************************
*   Copyright 2015 Florian Bauer, Michael Eischer, Christian Lobmeier,    *
*       Philipp Nordhus                                                   *
*   Robotics Erlangen e.V.                                                *
*   http://www.robotics-erlangen.de/                                      *
*   info@robotics-erlangen.de                                             *
*                                                                         *
*   This program is free software: you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation, either version 3 of the License, or     *
*   any later version.                                                    *
*                                                                         *
*   This program is distributed in the hope that it will be useful,       *
*   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
*   GNU General Public License for more details.                          *
*                                                                         *
*   You should have received a copy of the GNU General Public License     *
*   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
*************************************************************************]]

local vis = {}

local amun = amun
local Coordinates = require "../base/coordinates"


local gcolor = {}
local gisFilled = true

local ffi = require("ffi")
ffi.cdef[[
typedef struct { const unsigned char red, green, blue, alpha; } RGBA;
]]
local color_type_mt = {}
local color_type = ffi.metatype("RGBA", color_type_mt)

--- Joins rgba-value to a color.
-- Values from 0 to 255
-- @name fromRGBA
-- @param red number
-- @param green number
-- @param blue number
-- @param alpha number
-- @return table color
function vis.fromRGBA(red, green, blue, alpha)
	return color_type(red, green, blue, alpha)
end

--- Implements a red-yellow-green gradient
-- @name fromTemperature
-- @param value a normalized temperature [0, 1]
-- @param alpha the alpha value, default is 127
-- @return table color
function vis.fromTemperature(value, alpha)
	if value < 0 then
		error("vis temperature too low: " .. value)
	end
	if value > 1 then
		error("vis temperature too high: " .. value)
	end
	local red = 1
	local green = 1
	if value < 0.5 then
		red = 2 * value
	else
		green = 2 - 2 * value
	end
	return color_type(255 * red, 255 * green, 0, alpha or 127)
end

--- Modifies alpha value on a copy of the given color
-- @name setAlpha
-- @param color table - source color
-- @param alpha number - new alpha value
-- @return table - color with new alpha
function vis.setAlpha(color, alpha)
	local copy = table.copy(color)
	copy.alpha = alpha
	return copy
end

--- List of predefined colors.
-- with alpha = 255. Colors ending with half have alpha = 127.
-- @class table
-- @name colors
-- @field black (0,0,0)
-- @field white (255,255,255)
-- @field red (255,0,0)
-- @field green (0,255,0)
-- @field blue (0,0,255)
-- @field yellow (255,255,0)
-- @field pink (255,0,255)
-- @field turquoise (0,255,255)
-- @field orange (255, 127, 0)
-- @field magenta (255, 0, 127)
-- @field brown (127, 63, 0)
-- @field skyBlue (127, 191, 255)

-- @field blackHalf (0,0,0)
-- @field whiteHalf (255,255,255)
-- @field redHalf (255,0,0)
-- @field greenHalf (0,255,0)
-- @field blueHalf (0,0,255)
-- @field yellowHalf (255,255,0)
-- @field pinkHalf (255,0,255)
-- @field turquoiseHalf (0,255,255)
-- @field orangeHalf (255, 127, 0)
-- @field magentaHalf (255, 0, 127)
-- @field brownHalf (127, 63, 0)
-- @field skyBlueHalf (127, 191, 255)

vis.colors = {}

vis.colors.black = vis.fromRGBA(0, 0, 0, 255)
vis.colors.blackHalf = vis.fromRGBA(0, 0, 0, 127)
vis.colors.white = vis.fromRGBA(255, 255, 255, 255)
vis.colors.whiteHalf = vis.fromRGBA(255, 255, 255, 127)
vis.colors.grey = vis.fromRGBA(127, 127, 127, 255)
vis.colors.greyHalf = vis.fromRGBA(127, 127, 127, 127)

vis.colors.red = vis.fromRGBA(255, 0, 0, 255)
vis.colors.redHalf = vis.fromRGBA(255, 0, 0, 127)
vis.colors.green = vis.fromRGBA(0, 255, 0, 255)
vis.colors.greenHalf = vis.fromRGBA(0, 255, 0, 127)
vis.colors.blue = vis.fromRGBA(0, 0, 255, 255)
vis.colors.blueHalf = vis.fromRGBA(0, 0, 255, 127)

vis.colors.yellow = vis.fromRGBA(255, 255, 0, 255)
vis.colors.yellowHalf = vis.fromRGBA(255, 255, 0, 127)
vis.colors.pink = vis.fromRGBA(255, 0, 255, 255)
vis.colors.pinkHalf = vis.fromRGBA(255, 0, 255, 127)
vis.colors.turquoise = vis.fromRGBA(0, 255, 255, 255)
vis.colors.turquoiseHalf = vis.fromRGBA(0, 255, 255, 127)

vis.colors.orange = vis.fromRGBA(255, 127, 0, 255)
vis.colors.orangeHalf = vis.fromRGBA(255, 127, 0, 127)
vis.colors.magenta = vis.fromRGBA(255, 0, 127, 255)
vis.colors.magentaHalf = vis.fromRGBA(255, 0, 127, 127)
vis.colors.brown = vis.fromRGBA(127, 63, 0, 255)
vis.colors.brownHalf = vis.fromRGBA(127, 63, 0, 127)
vis.colors.skyBlue = vis.fromRGBA(127, 191, 255, 255)
vis.colors.skyBlueHalf = vis.fromRGBA(127, 191, 255, 127)

vis.colors.slate = vis.fromRGBA(112, 118, 144, 255)
vis.colors.slateHalf = vis.fromRGBA(112, 118, 144, 127)
vis.colors.orchid = vis.fromRGBA(218, 94, 224, 255)
vis.colors.orchidHalf = vis.fromRGBA(218, 94, 224, 127)
vis.colors.gold = vis.fromRGBA(239, 185, 15, 255)
vis.colors.goldHalf = vis.fromRGBA(239, 185, 15, 127)
vis.colors.mediumPurple = vis.fromRGBA(171, 130, 255, 255)
vis.colors.mediumPurpleHalf = vis.fromRGBA(171, 130, 255, 127)
vis.colors.darkPurple = vis.fromRGBA(93, 71, 139, 255)
vis.colors.darkPurpleHalf = vis.fromRGBA(93, 71, 139, 127)


--- Sets line and fill color.
-- If filled is true polygons and circles are filled using color.
-- @name setColor
-- @param color table
-- @param isFilled bool
function vis.setColor(color, isFilled)
	gcolor = color
	gisFilled = isFilled
end

--- Adds a circle.
-- If color is given use it instead of the global color and use the passed isFilled.
-- @name addCircle
-- @param name string - Visualization group
-- @param center Vector - center of the circle
-- @param radius number - radius of the circle
-- @param color table - color (optional)
-- @param isFilled bool - fill circle (optional)
function vis.addCircle(name, center, radius, color, isFilled, background, style, lineWidth)
	assert(radius, "missing radius parameter")
	vis.addCircleRaw(name, Coordinates.toGlobal(center), radius, color, isFilled, background, style, lineWidth)
end

--- Adds a circle. Requires global coordinates.
-- @name addCircleRaw
-- @see addCircle
function vis.addCircleRawGeneric(name, center, radius, color, isFilled, background, style, lineWidth)
	-- if color is set use passed isFilled
	if not color then
		isFilled = gisFilled
		color = gcolor
	end
	amun.addVisualization({
		name = name, pen = { color=color, style=style },
		brush = isFilled and color or nil, width = lineWidth or 0.01,
		circle = {p_x = center.x, p_y = center.y, radius = radius},
		background = background
	})
end

if amun.addVisualizationCircle then
	--- Adds a circle. Requires global coordinates.
	-- @name addCircleRaw
	-- @see addCircle
	function vis.addCircleRaw(name, center, radius, color, isFilled, background, style, lineWidth)
		if style ~= nil then
			return vis.addCircleRawGeneric(name, center, radius, color, isFilled, background, style, lineWidth)
		end

		-- if color is set use passed isFilled
		if not color then
			isFilled = gisFilled
			color = gcolor
		end
		amun.addVisualizationCircle(name, center.x, center.y, radius,
				color.red, color.green, color.blue, color.alpha,
				isFilled, background, lineWidth or 0.01)
	end
else
	vis.addCircleRaw = vis.addCircleRawGeneric
end

--- Adds a polygon.
-- If color is given use it instead of the global color and use the passed isFilled.
-- @name addPolygon
-- @param name string - Visualization group
-- @param points Vector[] - Points of the polygon
-- @param color table - color (optional)
-- @param isFilled bool - fill circle (optional)
function vis.addPolygon(name, points, color, isFilled, background, style)
	vis.addPolygonRaw(name, Coordinates.listToGlobal(points), color, isFilled, background, style)
end

--- Adds a polygon. Requires global coordinates.
-- @name addPolygonRaw
-- @see addPolygon
function vis.addPolygonRaw(name, points, color, isFilled, background, style)
	-- if color is set use passed isFilled
	if not color then
		isFilled = gisFilled
		color = gcolor
	end
	amun.addVisualization({
		name = name, pen = { color=color, style=style },
		brush = isFilled and color or nil, width = 0.01,
		polygon = {point = points},
		background = background
	})
end


--Paints an axis aligned rectangle
--@name addAxisAlignedRectangle
--@param name string - Visualization group
--@param corner1 Vector - One corner of the rectangle
--@param corner2 Vector - The other corner of the rectangle
--@param color table - see @addPolygon
--@param isFilled bool - see @addPolygon
--@param background - see @addPolygon
--@param style - see @addPolygon
function vis.addAxisAlignedRectangle(name, corner1, corner2, color, isFilled, background, style)
	local minX, minY, maxX, maxY
	minX = math.min(corner1.x, corner2.x)
	minY = math.min(corner1.y, corner2.y)
	maxX = math.max(corner1.x, corner2.x)
	maxY = math.max(corner1.y, corner2.y)
	local path = {}
	table.insert(path, Vector(minX, minY));
	table.insert(path, Vector(minX, maxY));
	table.insert(path, Vector(maxX, maxY));
	table.insert(path, Vector(maxX, minY));
	vis.addPolygon(name, path, color, isFilled, background, style)
end

--- Paints a Pizza where everything outside of [startAngle, endAngle] is filled
--- The shape of the pizza is approximated by a regular hexagon
-- @param name string - Name of the pizza
-- @param center Vectos - center point of the pizza
-- @param radius number - radius of the pizza
-- @param startAngle number - the starting angle of the missing pizza piece
-- @param endAngle number - the end angle of the missing pizza piece
local N_corners = 25
function vis.addPizza(name, center, radius, startAngle, endAngle, color, isFilled, background, style)
	local points = {center + Vector.fromAngle(startAngle)*radius, center, center + Vector.fromAngle(endAngle)*radius}
	if (startAngle - endAngle)%(2*math.pi) < 2*math.pi/N_corners then
		vis.addPolygon(name, points, color, isFilled, background, style)
	else
		local wStart = math.ceil(N_corners*endAngle/(2*math.pi))
		local wEnd = math.floor(N_corners*startAngle/(2*math.pi))
		if wEnd < wStart then
			wEnd = wEnd + N_corners
		end
		for w = wStart, wEnd do
			local angle = w*math.pi*2/N_corners
			table.insert(points, center + Vector.fromAngle(angle)*radius)
		end
		vis.addPolygon(name, points, color, isFilled, background, style)
	end
end

--- Adds a path.
-- If color is given use it instead of the global color and use the passed isFilled.
-- @name addPath
-- @param name string - Visualization group
-- @param points Vector[] - Points of the path
-- @param color table - line color (optional)
function vis.addPath(name, points, color, background, style, lineWidth)
	vis.addPathRaw(name, Coordinates.listToGlobal(points), color, background, style, lineWidth)
end

--- Adds a path. Requires global coordinates.
-- @name addPathRaw
-- @see addPath
function vis.addPathRaw(name, points, color, background, style, lineWidth)
	color = color or gcolor
	amun.addVisualization({
		name = name, pen = { color=color, style=style },
		width = lineWidth or 0.01,
		path = {point = points},
		background = background
	})
end

return vis

File: strategy/typescript/base/vis.ts
/**
 * @module vis
 * Provides functions to draw on the game field
 */

/**************************************************************************
*   Copyright 2018 Florian Bauer, Michael Eischer, Christian Lobmeier,    *
*       Philipp Nordhus, Andreas Wendler                                  *
*   Robotics Erlangen e.V.                                                *
*   http://www.robotics-erlangen.de/                                      *
*   info@robotics-erlangen.de                                             *
*                                                                         *
*   This program is free software: you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation, either version 3 of the License, ||     *
*   any later version.                                                    *
*                                                                         *
*   This program is distributed in the hope that it will be useful,       *
*   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*   MERCHANTABILITY || FITNESS FOR A PARTICULAR PURPOSE.  See the         *
*   GNU General Public License for more details.                          *
*                                                                         *
*   You should have received a copy of the GNU General Public License     *
*   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
**************************************************************************/
let amunLocal = amun;

import { Coordinates } from "base/coordinates";
import * as pb from "base/protobuf";
import { Position, Vector } from "base/vector";
import * as World from "base/world";


export class Color {
	public readonly red: number;
	public readonly green: number;
	public readonly blue: number;
	public readonly alpha: number;

	constructor(r: number, g: number, b: number, a: number) {
		this.red = r, this.green = g, this.blue = b, this.alpha = a;
	}

	setAlpha(a: number) {
		return new Color(this.red, this.green, this.blue, a);
	}
}

/**
 * Joins rgba-value to a color.
 * Values from 0 to 255
 */
export function fromRGBA(red: number, green: number, blue: number, alpha: number): Color {
	return new Color(red, green, blue, alpha);
}

/**
 * Implements a red-yellow-green gradient
 * @param value - a normalized temperature [0, 1]
 * @param alpha - the alpha value, default is 127
 */
export function fromTemperature(value: number, alpha: number = 127): Color {
	if (value < 0) {
		throw new Error(`vis temperature too low: ${value}`);
	}
	if (value > 1) {
		throw new Error(`vis temperature too high: ${value}`);
	}
	let red = 1;
	let green = 1;
	if (value < 0.5) {
		red = 2 * value;
	} else {
		green = 2 - 2 * value;
	}
	return new Color(255 * red, 255 * green, 0, alpha);
}

/**
 * Joins rgba-value to a color.
 * Values from 0 to 1
 */
export function fromHSVA(hue: number, saturation: number, value: number, alpha: number = 127): Color {
	let c = value * saturation;
	let x = c * (1 - Math.abs((hue * 6) % 2 - 1));
	let m = value - c;
	let r, g, b: number;
	r = c;
	g = x;
	b = 0;
	if (hue * 6 > 1) {
		r = x;
		g = c;
		b = 0;
	}
	if (hue * 6 > 2) {
		r = 0;
		g = c;
		b = x;
	}
	if (hue * 6 > 3) {
		r = 0;
		g = x;
		b = c;
	}
	if (hue * 6 > 4) {
		r = x;
		g = 0;
		b = c;
	}
	if (hue * 6 > 5) {
		r = c;
		g = 0;
		b = x;
	}
	return fromRGBA((r + m) * 255, (g + m) * 255, (b + m) * 255, alpha);
}

type Style = pb.amun.Pen.Style;

/**
 * List of predefined colors.
 *
 * Normals colors have alpha = 255.
 * Colors ending with half have alpha = 127.
 */
export const colors = {
	black: fromRGBA(0, 0, 0, 255),
	blackHalf: fromRGBA(0, 0, 0, 127),
	white: fromRGBA(255, 255, 255, 255),
	whiteHalf: fromRGBA(255, 255, 255, 127),
	whiteQuarter: fromRGBA(255, 255, 255, 64),
	grey: fromRGBA(127, 127, 127, 255),
	greyHalf: fromRGBA(127, 127, 127, 127),

	red: fromRGBA(255, 0, 0, 255),
	redHalf: fromRGBA(255, 0, 0, 127),
	green: fromRGBA(0, 255, 0, 255),
	greenHalf: fromRGBA(0, 255, 0, 127),
	blue: fromRGBA(0, 0, 255, 255),
	blueHalf: fromRGBA(0, 0, 255, 127),

	yellow: fromRGBA(255, 255, 0, 255),
	yellowHalf: fromRGBA(255, 255, 0, 127),
	pink: fromRGBA(255, 0, 255, 255),
	pinkHalf: fromRGBA(255, 0, 255, 127),
	cyan: fromRGBA(0, 255, 255, 255),
	cyanHalf: fromRGBA(0, 255, 255, 127),

	orange: fromRGBA(255, 127, 0, 255),
	orangeHalf: fromRGBA(255, 127, 0, 127),
	magenta: fromRGBA(255, 0, 127, 255),
	magentaHalf: fromRGBA(255, 0, 127, 127),
	brown: fromRGBA(127, 63, 0, 255),
	brownHalf: fromRGBA(127, 63, 0, 127),
	skyBlue: fromRGBA(127, 191, 255, 255),
	skyBlueHalf: fromRGBA(127, 191, 255, 127),

	slate: fromRGBA(112, 118, 144, 255),
	slateHalf: fromRGBA(112, 118, 144, 127),
	orchid: fromRGBA(218, 94, 224, 255),
	orchidHalf: fromRGBA(218, 94, 224, 127),
	gold: fromRGBA(239, 185, 15, 255),
	goldHalf: fromRGBA(239, 185, 15, 127),
	mediumPurple: fromRGBA(171, 130, 255, 255),
	mediumPurpleHalf: fromRGBA(171, 130, 255, 127),
	darkPurple: fromRGBA(93, 71, 139, 255),
	darkPurpleHalf: fromRGBA(93, 71, 139, 127),
};

let gcolor: Color = colors.black;
let gisFilled: boolean = true;

/**
 * Sets line and fill color.
 * If filled is true polygons and circles are filled using color.
 */
export function setColor(color: Color, isFilled: boolean) {
	gcolor = color;
	gisFilled = isFilled;
}

/**
 * Adds a circle.
 * If color is given use it instead of the global color and use the passed isFilled.
 * @param name - Visualization group
 * @param center - center of the circle
 * @param radius - radius of the circle
 * @param color - color (optional)
 * @param isFilled - fill circle (optional)
 */
export function addCircle(name: string, center: Position, radius: number, color?: Color,
		isFilled?: boolean, background?: boolean, style?: Style, lineWidth?: number) {
	addCircleRaw(name, Coordinates.toGlobal(center), radius, color, isFilled, background, style, lineWidth);
}

/**
 * Adds a circle. Requires global coordinates.
 * @see addCircle
 */
export function addCircleRaw(name: string, center: Position, radius: number, color?: Color,
		isFilled: boolean = false, background: boolean = false, style?: Style, lineWidth: number = 0.01) {
	// if color is set use passed isFilled
	if (color == undefined) {
		isFilled = gisFilled;
		color = gcolor;
	}
	if (style != undefined) {
		// style is not supported by the fast specialized version
		let brush: Color | undefined;
		if (isFilled) {
			brush = color;
		}
		let t: pb.amun.Visualization = {
			name: name, pen: { color: color, style: style },
			brush: brush, width: lineWidth,
			circle: { p_x: center.x, p_y: center.y, radius: radius },
			background: background
		};
		amunLocal.addVisualization(t);
	} else {
		amunLocal.addCircleSimple(name, center.x, center.y, radius, color.red,
			color.green, color.blue, color.alpha, isFilled, background, lineWidth);
	}
}

/**
 * Adds a polygon.
 * If color is given use it instead of the global color and use the passed isFilled.
 * @param name - Visualization group
 * @param points - Points of the polygon
 * @param color - color (optional)
 * @param isFilled - fill circle (optional)
 */
export function addPolygon(name: string, points: Position[], color?: Color,
		isFilled?: boolean, background?: boolean, style?: Style) {
	addPolygonRaw(name, Coordinates.listToGlobal(points), color, isFilled, background, style);
}

/**
 * Adds a polygon. Requires global coordinates.
 * @see addPolygon
 */
export function addPolygonRaw(name: string, points: Position[], color?: Color,
		isFilled: boolean = false, background: boolean = false, style?: Style) {
	// if color is set use passed isFilled
	if (color == undefined) {
		isFilled = gisFilled;
		color = gcolor;
	}
	if (style != undefined) {
		let brush: Color | undefined;
		if (isFilled) {
			brush = color;
		}
		amunLocal.addVisualization({
			name: name, pen: { color: color, style: style },
			brush: brush, width: 0.01,
			polygon: { point: points },
			background: background
		});
	} else {
		let pointArray: number[] = [];
		for (let pos of points) {
			pointArray.push(pos.x);
			pointArray.push(pos.y);
		}
		amunLocal.addPolygonSimple(name, color.red, color.green, color.blue, color.alpha,
			isFilled, background, pointArray);
	}
}


/**
 * Paints an axis aligned rectangle
 * @name addAxisAlignedRectangle
 * @param name - Visualization group
 * @param corner1 - One corner of the rectangle
 * @param corner2 - The other corner of the rectangle
 * @param color - see @addPolygon
 * @param isFilled - see @addPolygon
 * @param background - see @addPolygon
 * @param style - see @addPolygon
 */
export function addAxisAlignedRectangle(name: string, corner1: Position, corner2: Position,
		color?: Color, isFilled?: boolean, background?: boolean, style?: Style) {
	let minX, minY, maxX, maxY;
	minX = Math.min(corner1.x, corner2.x);
	minY = Math.min(corner1.y, corner2.y);
	maxX = Math.max(corner1.x, corner2.x);
	maxY = Math.max(corner1.y, corner2.y);
	let path: Position[] = [];
	path[0] = new Vector(minX, minY);
	path[1] = new Vector(minX, maxY);
	path[2] = new Vector(maxX, maxY);
	path[3] = new Vector(maxX, minY);
	addPolygon(name, path, color, isFilled, background, style);
}

/**
 * Paints a Pizza where everything outside of [startAngle, endAngle] is filled
 * The shape of the pizza is approximated by a regular hexagon
 * @param name - the pizza
 * @param center - center point of the pizza
 * @param radius - radius of the pizza
 * @param startAngle - the starting angle of the missing pizza piece
 * @param endAngle - the end angle of the missing pizza piece
 */
const N_CORNERS = 25;
export function addPizza(name: string, center: Position, radius: number,
		startAngle: number, endAngle: number, color?: Color, isFilled?: boolean, background?: boolean, style?: Style) {
	let points = [center + Vector.fromPolar(startAngle, radius), center, center + Vector.fromPolar(endAngle, radius)];
	if ((startAngle - endAngle) % (2 * Math.PI) < 2 * Math.PI / N_CORNERS) {
		addPolygon(name, points, color, isFilled, background, style);
	} else {
		let wStart = Math.ceil(N_CORNERS * endAngle / (2 * Math.PI));
		let wEnd = Math.floor(N_CORNERS * startAngle / (2 * Math.PI));
		if (wEnd < wStart) {
			wEnd = wEnd + N_CORNERS;
		}
		for (let w = wStart; w < wEnd; w++) {
			let angle = w * Math.PI * 2 / N_CORNERS;
			points.push(center + Vector.fromPolar(angle, radius));
		}
		addPolygon(name, points, color, isFilled, background, style);
	}
}

/**
 * Adds a path.
 * If color is given use it instead of the global color and use the passed isFilled.
 * @param name - Visualization group
 * @param points - Points of the path
 * @param color - line color (optional)
 */
export function addPath(name: string, points: Position[], color?: Color, background?: boolean, style?: Style, lineWidth?: number) {
	addPathRaw(name, Coordinates.listToGlobal(points), color, background, style, lineWidth);
}

/**
 * Adds a path. Requires global coordinates.
 * @see addPath
 */
export function addPathRaw(name: string, points: Position[], color: Color = gcolor, background: boolean = false,
		style?: Style, lineWidth: number = 0.01) {
	if (style != undefined) {
		amunLocal.addVisualization({
			name: name, pen: { color: color, style: style },
			width: lineWidth,
			path: { point: points },
			background: background
		});
	} else {
		if (amun.SUPPORTS_EFFICIENT_PATHVIS) {
			let allData = new Float32Array(6 + 2 * points.length);
			allData[0] = color.red;
			allData[1] = color.green;
			allData[2] = color.blue;
			allData[3] = color.alpha;
			allData[4] = lineWidth;
			allData[5] = background ? 1 : 0;
			for (let i = 0; i < points.length; i++) {
				allData[6 + i * 2] = points[i].x;
				allData[6 + i * 2 + 1] = points[i].y;
			}
			amunLocal.addPathSimple(name, allData);
		} else {
			let pointArray: number[] = [];
			for (let pos of points) {
				pointArray.push(pos.x);
				pointArray.push(pos.y);
			}
			amunLocal.addPathSimple(name, color.red, color.green, color.blue, color.alpha, lineWidth, background, pointArray);
		}
	}
}

export function addFieldVisualization(name: string, f: (pos: Vector) => Color, pixelWidth: number, pixelHeight: number,
		drawCornerTL?: Vector, drawCornerBR?: Vector) {

	if (!drawCornerTL) {
		drawCornerTL = new Vector(-World.Geometry.FieldWidthHalf - World.Geometry.BoundaryWidth,
			-World.Geometry.FieldHeightHalf - World.Geometry.BoundaryWidth);
	} else {
		drawCornerTL = Coordinates.toGlobal(drawCornerTL);
	}
	if (!drawCornerBR) {
		drawCornerBR = new Vector(World.Geometry.FieldWidthHalf + World.Geometry.BoundaryWidth,
			World.Geometry.FieldHeightHalf + World.Geometry.BoundaryWidth);
	} else {
		drawCornerBR = Coordinates.toGlobal(drawCornerBR);
	}
	let data = new Uint8Array(pixelWidth * pixelHeight * 4);
	const fieldSize = drawCornerBR - drawCornerTL;
	for (let y = 0; y < pixelHeight; y++) {
		for (let x = 0; x < pixelWidth; x++) {
			const pos = new Vector((x + 0.5) / pixelWidth * fieldSize.x,
									(y + 0.5) / pixelHeight * fieldSize.y) + drawCornerTL;
			const color = f(Coordinates.toLocal(pos));
			const baseIndex = (y * pixelWidth + x) * 4;
			data.set([color.blue, color.green, color.red, color.alpha], baseIndex);
		}
	}
	amunLocal.addVisualization({ name: name, image: {
		width: pixelWidth,
		height: pixelHeight,
		data: data,
		draw_area: {
			topleft: {
				x: drawCornerTL.x,
				y: drawCornerTL.y
			},
			bottomright: {
				x: drawCornerBR.x,
				y: drawCornerBR.y
			}
		}
	} } as any);
}
